### PR TITLE
bpo-30773: Fix ag_running; prohibit running athrow/asend/aclose in parallel

### PR DESCRIPTION
**Pull request by 1st1**
https://bugs.python.org/issue30773

### CODE REVIEW

File: Include/genobject.h [hunk 1]
```diff
@@ -78,6 +78,8 @@ typedef struct {
     /* Flag is set to 1 when aclose() is called for the first time, or
        when a StopAsyncIteration exception is raised. */
     int ag_closed;
+
+    int ag_running_async;
```
Comment (dimaqq) [inline Include/genobject.h:83]: Considering that it's a top-level core data structure, I think a comment for this field is warranted.

File: Lib/asyncio/base_events.py [hunk 2]
```diff
@@ -490,7 +490,7 @@ def _asyncgen_firstiter_hook(self, agen):
         self._asyncgens.clear()
 
         results = await tasks.gather(
-            *[ag.aclose() for ag in closing_agens],
+            *[ag.aclose() for ag in closing_agens if not ag.ag_running],
```
Comment (dimaqq) [inline Lib/asyncio/base_events.py]: So, if some generator is running (not that it should, I think, at this point), is going to be left behind?
Or is perhaps assumed that closing one generator might set another as running?

File: Misc/NEWS.d/next/Core and Builtins/2018-06-07-01-01-20.bpo-30773.C31rVE.rst [hunk 3]
```diff
@@ -0,0 +1,2 @@
+Prohibit parallel running of aclose() / asend() / athrow(). Fix ag_running
+to reflect the actual running sttaus of the AG.
```
Comment (dimaqq) [inline Misc/NEWS.d/next/Core and Builtins/2018-06-07-01-01-20.bpo-30773.C31rVE.rst]: typo

File: Lib/asyncio/base_events.py (hunk 2 above)
Comment (1st1) [reply Lib/asyncio/base_events.py]: Yeah, this change should be reverted (i didn't intend this to be committed)

File: Include/genobject.h (hunk 1 above)
Comment (1st1) [reply Include/genobject.h:83]: Yep, i'll add a comment later. This PR is wip.

**Comment by asvetlov** [comment]
Could you resolve conflicts please?
I'd like to checkout your branch and play with it locally.

**Comment by 1st1** [comment]
> Could you resolve conflicts please?
I'd like to checkout your branch and play with it locally.

Done!

File: Misc/NEWS.d/next/Core and Builtins/2018-06-07-01-01-20.bpo-30773.C31rVE.rst (hunk 3 above)
Comment (asvetlov) [inline Misc/NEWS.d/next/Core and Builtins/2018-06-07-01-01-20.bpo-30773.C31rVE.rst]:
```suggestion
to reflect the actual running status of the AG.
```

File: Lib/asyncio/base_events.py (hunk 2 above)
Comment (asvetlov) [reply Lib/asyncio/base_events.py]: Is it still relevant?

**Comment by asvetlov** [comment]
@1st1 the PR looks good after fixing a couple comments

**Comment by 1st1** [comment]
> @1st1 the PR looks good after fixing a couple comments

Thanks, will fix it and commit soon.

File: Lib/asyncio/base_events.py (hunk 2 above)
Comment (Hina-creator) [reply Lib/asyncio/base_events.py]: > So, if some generator is running (not that it should, I think, at this point), is going to be left behind?
> 
> Or is perhaps assumed that closing one generator might set another as running?

**Pull request merged**